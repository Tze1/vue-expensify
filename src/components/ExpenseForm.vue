<template>
  <div class="expenseform component">
    <p v-if="!!errorMessage" class="expenseform-error">{{ errorMessage }}</p>
    <form @submit="onSubmit">
      <Datepicker
        :clear-button="false"
        @selected="onDateChange"
        calendar-class="expenseform-datepicker"
        format="MM/dd/yyyy"
        input-class="datepicker-input"
        name="createdAt"
        :value="date"
      />
      <input
        class="expenseform-description"
        type="text"
        name="description"
        placeholder="Description"
        autoFocus
        v-model="description"
      />
      <input
        class="expenseform-amount"
        type="text"
        name="amount"
        placeholder="Amount"
        v-model="amount"
      />
      <textarea
        class="expenseform-note"
        name="note"
        placeholder="Add a note.  (optional)"
        v-model="note"
      />
      <button class="expenseform-submit button" type="submit">
        {{ expense ? 'Update' : 'Save' }} Expense
      </button>
    </form>
  </div>
</template>

<script>
  import Datepicker from 'vuejs-datepicker';
  import moment from 'moment';

  export default {
    name: 'ExpenseForm',
    data () {
      return {
        date: this.expense ?
          moment(this.expense.createdAt).toDate() :
          moment().toDate(),
        description: this.expense ?
          this.expense.description :
          '',
        amount: this.expense ?
          (this.expense.amount / 100).toString() :
          '',
        note: this.expense ?
          this.expense.note :
          '',
        errorMessage: '',
      }
    },
    props: {
      submitExpense: {
        type: Function,
        required: true,
      },
      expense: {
        type: Object,
        required: false,
      },
    },
    watch: {
      expense: function () {
        /* After any manual page-refresh, if an expense does get passed in for edit, populate form-fields (set data) with expense-info. */
        const expense = this.expense;
        if (expense) {
          this.date = moment(expense.createdAt).toDate();
          this.description = expense.description;
          this.amount = (expense.amount / 100).toString();
          this.note = expense.note;
        }
      },
    },
    methods: {
      onDateChange (e) {
        this.date = e;
      },
      onSubmit (e) {
        const vm = this;
        e.preventDefault();  // don't navigate!

        if (!!vm.description && !!vm.amount) {
          // Clear error message and call parent submit method (from props).
          vm.errorMessage = '';
          vm.submitExpense({
            createdAt: moment(vm.date.getTime()).valueOf(),
            description: vm.description,
            amount: parseFloat(vm.amount * 100),
            note: vm.note,
          });
        } else {
          // Set error message.
          vm.errorMessage = 'ERROR: Missing description or amount.';
        }
      },
    },
    components: {
      Datepicker,
    },
  };
</script>

<style scoped lang="scss">
  @import '../styles/app.scss';

  .expenseform.component {
    form {
      > * {
        display: block;
        margin-bottom: $space-sm;
      }

      > input,
      > textarea {
        border: 0;
        padding: $space-xs;
        width: calc(100% - #{$space-lg});

        @include mq("tablet") {
          min-width: 23.75rem;
          width: auto;
        }

        &.expenseform-amount {
          text-align: right;
        }
      }  // > .input, > .textarea
    }  // form
  }  // .expenseform.component
</style>
