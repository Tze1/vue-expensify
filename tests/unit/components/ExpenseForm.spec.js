import { mount } from '@vue/test-utils';
import ExpenseForm from '@/components/ExpenseForm';
import testExpenses from '../../fixtures/testExpenses';
import moment from 'moment';
import Datepicker from 'vuejs-datepicker';

describe('ExpenseForm.vue', () => {
  const onDateChangeStub = jest.fn();
  const submitExpenseStub = jest.fn();
  const addExpensePropsData = {
    submitExpense: submitExpenseStub,
    expense: undefined,
  };
  const testExpense = { ...testExpenses[2] };
  const editExpensePropsData = {
    submitExpense: submitExpenseStub,
    expense: { ...testExpense },
  };
  const wrapperMethods = {
    onDateChange: onDateChangeStub,
  };

  it('renders ExpenseForm properly for add-expense', () => {
    const wrapper = mount(ExpenseForm, {
      data () {
        return {
          date: moment().toDate(),
          description: '',
          amount: '',
          note: '',
          errorMessage: '',
        }
      },
      propsData: addExpensePropsData,
    });

    wrapper.setMethods(wrapperMethods);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders ExpenseForm properly for edit-expense', () => {
    const wrapper = mount(ExpenseForm, {
      data () {
        return {
          date: moment().toDate(),
          description: '',
          amount: '',
          note: '',
          errorMessage: '',
        }
      },
      propsData: editExpensePropsData,
    });

    wrapper.setMethods(wrapperMethods);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.props().expense.description).toEqual(testExpense.description);
  });

  it('renders error-message on invalid submission', () => {
    const wrapper = mount(ExpenseForm, {
      data () {
        return {
          date: moment().toDate(),
          description: '',
          amount: '',
          note: '',
          errorMessage: '',
        }
      },
      propsData: addExpensePropsData,
    });

    wrapper.setMethods(wrapperMethods);

    expect(wrapper).toMatchSnapshot();

    wrapper.find('form').trigger('submit');

    expect(wrapper.vm.errorMessage.length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
  });

  it('calls parent submitExpense properly on valid submission', () => {
    const wrapper = mount(ExpenseForm, {
      data () {
        return {
          date: moment(testExpense.createdAt).toDate(),
          description: testExpense.description,
          amount: (testExpense.amount / 100).toString(),
          note: testExpense.note,
          errorMessage: '',
        }
      },
      propsData: editExpensePropsData,
    });

    wrapper.setMethods(wrapperMethods);

    expect(wrapper).toMatchSnapshot();
    
    wrapper.find('form').trigger('submit');

    expect(submitExpenseStub).toHaveBeenCalledWith({
      createdAt: testExpense.createdAt,
      description: testExpense.description,
      amount: testExpense.amount,
      note: testExpense.note,
    });
  });
});
