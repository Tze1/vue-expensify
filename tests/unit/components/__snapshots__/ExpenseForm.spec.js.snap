// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseForm.vue calls parent submitExpense properly on valid submission 1`] = `
<div class="expenseform component">
  <!---->
  <form>
    <div class="vdp-datepicker">
      <div class="">
        <!----> <input type="text" name="createdAt" readonly="readonly" autocomplete="off" class="datepicker-input">
        <!---->
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="day__month_btn up">Dec 2018</span> <span class="next">&gt;</span></header>
        <div class=""><span class="cell day-header">Sun</span><span class="cell day-header">Mon</span><span class="cell day-header">Tue</span><span class="cell day-header">Wed</span><span class="cell day-header">Thu</span><span class="cell day-header">Fri</span><span class="cell day-header">Sat</span> <span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day weekend sat">1</span><span class="cell day weekend sun">2</span><span class="cell day">3</span><span class="cell day">4</span><span class="cell day">5</span><span class="cell day">6</span><span class="cell day">7</span><span class="cell day weekend sat">8</span><span class="cell day today weekend sun">9</span><span class="cell day">10</span><span class="cell day">11</span><span class="cell day">12</span><span class="cell day">13</span><span class="cell day">14</span><span class="cell day weekend sat">15</span><span class="cell day weekend sun">16</span><span class="cell day">17</span><span class="cell day">18</span><span class="cell day">19</span><span class="cell day">20</span><span class="cell day">21</span><span class="cell day weekend sat">22</span><span class="cell day weekend sun">23</span><span class="cell day">24</span><span class="cell day">25</span><span class="cell day">26</span><span class="cell day">27</span><span class="cell day">28</span><span class="cell day weekend sat">29</span><span class="cell day weekend sun">30</span><span class="cell day">31</span></div>
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="month__year_btn up">2018</span> <span class="next">&gt;</span></header> <span class="cell month">January</span><span class="cell month">February</span><span class="cell month">March</span><span class="cell month">April</span><span class="cell month">May</span><span class="cell month">June</span><span class="cell month">July</span><span class="cell month">August</span><span class="cell month">September</span><span class="cell month">October</span><span class="cell month">November</span><span class="cell month">December</span>
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span>2010 - 2019</span> <span class="next">&gt;</span></header> <span class="cell year">2010</span><span class="cell year">2011</span><span class="cell year">2012</span><span class="cell year">2013</span><span class="cell year">2014</span><span class="cell year">2015</span><span class="cell year">2016</span><span class="cell year">2017</span><span class="cell year">2018</span><span class="cell year">2019</span>
      </div>
    </div> <input type="text" name="description" placeholder="Description" autofocus="" class="expenseform-description"> <input type="text" name="amount" placeholder="Amount" class="expenseform-amount"> <textarea name="note" placeholder="Add a note.  (optional)" class="expenseform-note"></textarea> <button type="submit" class="expenseform-submit button">
      Update Expense
    </button>
  </form>
</div>
`;

exports[`ExpenseForm.vue renders ExpenseForm properly for add-expense 1`] = `
<div class="expenseform component">
  <!---->
  <form>
    <div class="vdp-datepicker">
      <div class="">
        <!----> <input type="text" name="createdAt" readonly="readonly" autocomplete="off" class="datepicker-input">
        <!---->
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="day__month_btn up">Dec 2018</span> <span class="next">&gt;</span></header>
        <div class=""><span class="cell day-header">Sun</span><span class="cell day-header">Mon</span><span class="cell day-header">Tue</span><span class="cell day-header">Wed</span><span class="cell day-header">Thu</span><span class="cell day-header">Fri</span><span class="cell day-header">Sat</span> <span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day weekend sat">1</span><span class="cell day weekend sun">2</span><span class="cell day">3</span><span class="cell day">4</span><span class="cell day">5</span><span class="cell day">6</span><span class="cell day">7</span><span class="cell day weekend sat">8</span><span class="cell day today weekend sun">9</span><span class="cell day">10</span><span class="cell day">11</span><span class="cell day">12</span><span class="cell day">13</span><span class="cell day">14</span><span class="cell day weekend sat">15</span><span class="cell day weekend sun">16</span><span class="cell day">17</span><span class="cell day">18</span><span class="cell day">19</span><span class="cell day">20</span><span class="cell day">21</span><span class="cell day weekend sat">22</span><span class="cell day weekend sun">23</span><span class="cell day">24</span><span class="cell day">25</span><span class="cell day">26</span><span class="cell day">27</span><span class="cell day">28</span><span class="cell day weekend sat">29</span><span class="cell day weekend sun">30</span><span class="cell day">31</span></div>
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="month__year_btn up">2018</span> <span class="next">&gt;</span></header> <span class="cell month">January</span><span class="cell month">February</span><span class="cell month">March</span><span class="cell month">April</span><span class="cell month">May</span><span class="cell month">June</span><span class="cell month">July</span><span class="cell month">August</span><span class="cell month">September</span><span class="cell month">October</span><span class="cell month">November</span><span class="cell month">December</span>
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span>2010 - 2019</span> <span class="next">&gt;</span></header> <span class="cell year">2010</span><span class="cell year">2011</span><span class="cell year">2012</span><span class="cell year">2013</span><span class="cell year">2014</span><span class="cell year">2015</span><span class="cell year">2016</span><span class="cell year">2017</span><span class="cell year">2018</span><span class="cell year">2019</span>
      </div>
    </div> <input type="text" name="description" placeholder="Description" autofocus="" class="expenseform-description"> <input type="text" name="amount" placeholder="Amount" class="expenseform-amount"> <textarea name="note" placeholder="Add a note.  (optional)" class="expenseform-note"></textarea> <button type="submit" class="expenseform-submit button">
      Save Expense
    </button>
  </form>
</div>
`;

exports[`ExpenseForm.vue renders ExpenseForm properly for edit-expense 1`] = `
<div class="expenseform component">
  <!---->
  <form>
    <div class="vdp-datepicker">
      <div class="">
        <!----> <input type="text" name="createdAt" readonly="readonly" autocomplete="off" class="datepicker-input">
        <!---->
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="day__month_btn up">Dec 2018</span> <span class="next">&gt;</span></header>
        <div class=""><span class="cell day-header">Sun</span><span class="cell day-header">Mon</span><span class="cell day-header">Tue</span><span class="cell day-header">Wed</span><span class="cell day-header">Thu</span><span class="cell day-header">Fri</span><span class="cell day-header">Sat</span> <span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day weekend sat">1</span><span class="cell day weekend sun">2</span><span class="cell day">3</span><span class="cell day">4</span><span class="cell day">5</span><span class="cell day">6</span><span class="cell day">7</span><span class="cell day weekend sat">8</span><span class="cell day today weekend sun">9</span><span class="cell day">10</span><span class="cell day">11</span><span class="cell day">12</span><span class="cell day">13</span><span class="cell day">14</span><span class="cell day weekend sat">15</span><span class="cell day weekend sun">16</span><span class="cell day">17</span><span class="cell day">18</span><span class="cell day">19</span><span class="cell day">20</span><span class="cell day">21</span><span class="cell day weekend sat">22</span><span class="cell day weekend sun">23</span><span class="cell day">24</span><span class="cell day">25</span><span class="cell day">26</span><span class="cell day">27</span><span class="cell day">28</span><span class="cell day weekend sat">29</span><span class="cell day weekend sun">30</span><span class="cell day">31</span></div>
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="month__year_btn up">2018</span> <span class="next">&gt;</span></header> <span class="cell month">January</span><span class="cell month">February</span><span class="cell month">March</span><span class="cell month">April</span><span class="cell month">May</span><span class="cell month">June</span><span class="cell month">July</span><span class="cell month">August</span><span class="cell month">September</span><span class="cell month">October</span><span class="cell month">November</span><span class="cell month">December</span>
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span>2010 - 2019</span> <span class="next">&gt;</span></header> <span class="cell year">2010</span><span class="cell year">2011</span><span class="cell year">2012</span><span class="cell year">2013</span><span class="cell year">2014</span><span class="cell year">2015</span><span class="cell year">2016</span><span class="cell year">2017</span><span class="cell year">2018</span><span class="cell year">2019</span>
      </div>
    </div> <input type="text" name="description" placeholder="Description" autofocus="" class="expenseform-description"> <input type="text" name="amount" placeholder="Amount" class="expenseform-amount"> <textarea name="note" placeholder="Add a note.  (optional)" class="expenseform-note"></textarea> <button type="submit" class="expenseform-submit button">
      Update Expense
    </button>
  </form>
</div>
`;

exports[`ExpenseForm.vue renders error-message on invalid submission 1`] = `
<div class="expenseform component">
  <!---->
  <form>
    <div class="vdp-datepicker">
      <div class="">
        <!----> <input type="text" name="createdAt" readonly="readonly" autocomplete="off" class="datepicker-input">
        <!---->
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="day__month_btn up">Dec 2018</span> <span class="next">&gt;</span></header>
        <div class=""><span class="cell day-header">Sun</span><span class="cell day-header">Mon</span><span class="cell day-header">Tue</span><span class="cell day-header">Wed</span><span class="cell day-header">Thu</span><span class="cell day-header">Fri</span><span class="cell day-header">Sat</span> <span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day weekend sat">1</span><span class="cell day weekend sun">2</span><span class="cell day">3</span><span class="cell day">4</span><span class="cell day">5</span><span class="cell day">6</span><span class="cell day">7</span><span class="cell day weekend sat">8</span><span class="cell day today weekend sun">9</span><span class="cell day">10</span><span class="cell day">11</span><span class="cell day">12</span><span class="cell day">13</span><span class="cell day">14</span><span class="cell day weekend sat">15</span><span class="cell day weekend sun">16</span><span class="cell day">17</span><span class="cell day">18</span><span class="cell day">19</span><span class="cell day">20</span><span class="cell day">21</span><span class="cell day weekend sat">22</span><span class="cell day weekend sun">23</span><span class="cell day">24</span><span class="cell day">25</span><span class="cell day">26</span><span class="cell day">27</span><span class="cell day">28</span><span class="cell day weekend sat">29</span><span class="cell day weekend sun">30</span><span class="cell day">31</span></div>
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="month__year_btn up">2018</span> <span class="next">&gt;</span></header> <span class="cell month">January</span><span class="cell month">February</span><span class="cell month">March</span><span class="cell month">April</span><span class="cell month">May</span><span class="cell month">June</span><span class="cell month">July</span><span class="cell month">August</span><span class="cell month">September</span><span class="cell month">October</span><span class="cell month">November</span><span class="cell month">December</span>
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span>2010 - 2019</span> <span class="next">&gt;</span></header> <span class="cell year">2010</span><span class="cell year">2011</span><span class="cell year">2012</span><span class="cell year">2013</span><span class="cell year">2014</span><span class="cell year">2015</span><span class="cell year">2016</span><span class="cell year">2017</span><span class="cell year">2018</span><span class="cell year">2019</span>
      </div>
    </div> <input type="text" name="description" placeholder="Description" autofocus="" class="expenseform-description"> <input type="text" name="amount" placeholder="Amount" class="expenseform-amount"> <textarea name="note" placeholder="Add a note.  (optional)" class="expenseform-note"></textarea> <button type="submit" class="expenseform-submit button">
      Save Expense
    </button>
  </form>
</div>
`;

exports[`ExpenseForm.vue renders error-message on invalid submission 2`] = `
<div class="expenseform component">
  <p class="expenseform-error">ERROR: Missing description or amount.</p>
  <form>
    <div class="vdp-datepicker">
      <div class="">
        <!----> <input type="text" name="createdAt" readonly="readonly" autocomplete="off" class="datepicker-input">
        <!---->
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="day__month_btn up">Dec 2018</span> <span class="next">&gt;</span></header>
        <div class=""><span class="cell day-header">Sun</span><span class="cell day-header">Mon</span><span class="cell day-header">Tue</span><span class="cell day-header">Wed</span><span class="cell day-header">Thu</span><span class="cell day-header">Fri</span><span class="cell day-header">Sat</span> <span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day weekend sat">1</span><span class="cell day weekend sun">2</span><span class="cell day">3</span><span class="cell day">4</span><span class="cell day">5</span><span class="cell day">6</span><span class="cell day">7</span><span class="cell day weekend sat">8</span><span class="cell day today weekend sun">9</span><span class="cell day">10</span><span class="cell day">11</span><span class="cell day">12</span><span class="cell day">13</span><span class="cell day">14</span><span class="cell day weekend sat">15</span><span class="cell day weekend sun">16</span><span class="cell day">17</span><span class="cell day">18</span><span class="cell day">19</span><span class="cell day">20</span><span class="cell day">21</span><span class="cell day weekend sat">22</span><span class="cell day weekend sun">23</span><span class="cell day">24</span><span class="cell day">25</span><span class="cell day">26</span><span class="cell day">27</span><span class="cell day">28</span><span class="cell day weekend sat">29</span><span class="cell day weekend sun">30</span><span class="cell day">31</span></div>
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="month__year_btn up">2018</span> <span class="next">&gt;</span></header> <span class="cell month">January</span><span class="cell month">February</span><span class="cell month">March</span><span class="cell month">April</span><span class="cell month">May</span><span class="cell month">June</span><span class="cell month">July</span><span class="cell month">August</span><span class="cell month">September</span><span class="cell month">October</span><span class="cell month">November</span><span class="cell month">December</span>
      </div>
      <div class="expenseform-datepicker vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span>2010 - 2019</span> <span class="next">&gt;</span></header> <span class="cell year">2010</span><span class="cell year">2011</span><span class="cell year">2012</span><span class="cell year">2013</span><span class="cell year">2014</span><span class="cell year">2015</span><span class="cell year">2016</span><span class="cell year">2017</span><span class="cell year">2018</span><span class="cell year">2019</span>
      </div>
    </div> <input type="text" name="description" placeholder="Description" autofocus="" class="expenseform-description"> <input type="text" name="amount" placeholder="Amount" class="expenseform-amount"> <textarea name="note" placeholder="Add a note.  (optional)" class="expenseform-note"></textarea> <button type="submit" class="expenseform-submit button">
      Save Expense
    </button>
  </form>
</div>
`;
